<a href="https://colab.research.google.com/github/HydrOpOtesJA99/A.I-learning/blob/main/11_%EC%95%84%EC%9D%B4%EB%A6%AC%EC%8A%A4_%ED%92%88%EC%A2%85_%EB%B6%84%EB%A5%98_%EC%8B%A4%EC%8A%B5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 11_아이리스_품종_분류_실습



*   github csv url: https://raw.githubusercontent.com/blackdew/tensorflow1/master/csv/iris.csv



## 라이브러리 사용

In [1]:
# 라이브러리 사용
import tensorflow as tf
import pandas as pd

## 1. 과거의 데이터를 준비한다

In [2]:
# 과거의 데이터를 준비
파일경로 = 'https://raw.githubusercontent.com/blackdew/tensorflow1/master/csv/iris.csv'
아이리스 = pd.read_csv(파일경로)
아이리스.head()

꽃잎길이  꽃잎폭  꽃받침길이  꽃받침폭      품종
0   5.1  3.5    1.4   0.2  setosa
1   4.9  3.0    1.4   0.2  setosa
2   4.7  3.2    1.3   0.2  setosa
3   4.6  3.1    1.5   0.2  setosa
4   5.0  3.6    1.4   0.2  setosa

In [11]:
# 원핫인코딩 -> 변수형 데이터를 실수형 데이터로 변환
인코딩 = pd.get_dummies(아이리스)

인코딩.head()

꽃잎길이  꽃잎폭  꽃받침길이  꽃받침폭  품종_setosa  품종_versicolor  품종_virginica
0   5.1  3.5    1.4   0.2          1              0             0
1   4.9  3.0    1.4   0.2          1              0             0
2   4.7  3.2    1.3   0.2          1              0             0
3   4.6  3.1    1.5   0.2          1              0             0
4   5.0  3.6    1.4   0.2          1              0             0

In [13]:
print(인코딩.columns)

Index(['꽃잎길이', '꽃잎폭', '꽃받침길이', '꽃받침폭', '품종_setosa', '품종_versicolor',
       '품종_virginica'],
      dtype='object')


In [16]:
# 독립변수, 종속변수 구분

독립 = 인코딩[['꽃잎길이', '꽃잎폭', '꽃받침길이', '꽃받침폭']] # 대괄호, 대괄호 ([[)로 해주야 함
종속 = 인코딩[['품종_setosa', '품종_versicolor', '품종_virginica']]
print(독립.shape, 종속.shape)

(150, 4) (150, 3)


## 2. 모델의 구조를 만든다.

In [17]:
X = tf.keras.layers.Input(shape=[4])
Y = tf.keras.layers.Dense(3, activation='softmax')(X)
model = tf.keras.models.Model(X,Y)
model.compile(loss='categorical_crossentropy',
              metrics='accuracy')

## 3. 데이터로 모델을 학습(FIT)시킨다.

In [21]:
# 모델을 학습
model.fit(독립, 종속, epochs=2000)
# model.fit(독립, 종속, epochs=2000, verbose=0) verbose=0: 학습하는 결과가 창에 뜨지 않음

Epoch 1/10
5/5 [==============================] - 0s 3ms/step - loss: 0.1443 - accuracy: 0.9733
Epoch 2/10
5/5 [==============================] - 0s 3ms/step - loss: 0.1439 - accuracy: 0.9733
Epoch 3/10
5/5 [==============================] - 0s 3ms/step - loss: 0.1437 - accuracy: 0.9733
Epoch 4/10
5/5 [==============================] - 0s 4ms/step - loss: 0.1438 - accuracy: 0.9733
Epoch 5/10
5/5 [==============================] - 0s 4ms/step - loss: 0.1434 - accuracy: 0.9733
Epoch 6/10
5/5 [==============================] - 0s 3ms/step - loss: 0.1442 - accuracy: 0.9733
Epoch 7/10
5/5 [==============================] - 0s 3ms/step - loss: 0.1434 - accuracy: 0.9733
Epoch 8/10
5/5 [==============================] - 0s 3ms/step - loss: 0.1432 - accuracy: 0.9733
Epoch 9/10
5/5 [==============================] - 0s 3ms/step - loss: 0.1437 - accuracy: 0.9733
Epoch 10/10
5/5 [==============================] - 0s 3ms/step - loss: 0.1434 - accuracy: 0.9733


## 4. 모델을 이용한다.

In [22]:
# 모델의 이용
model.predict(독립[0:5])

array([[9.99554455e-01, 4.45574638e-04, 1.37402667e-08],
       [9.97987390e-01, 2.01247539e-03, 1.47118485e-07],
       [9.99101996e-01, 8.98085302e-04, 5.75134216e-08],
       [9.97387111e-01, 2.61250115e-03, 2.97741849e-07],
       [9.99642611e-01, 3.57340381e-04, 1.11824665e-08]], dtype=float32)

In [23]:
print(종속[0:5])

   품종_setosa  품종_versicolor  품종_virginica
0          1              0             0
1          1              0             0
2          1              0             0
3          1              0             0
4          1              0             0


In [25]:
model.predict(독립[-5:]) # '-5:': 뒤에서 5개의 값을 말한다(파이썬 slicing)

array([[1.70975682e-07, 9.22614485e-02, 9.07738328e-01],
       [4.58195444e-07, 1.74210206e-01, 8.25789273e-01],
       [8.28449515e-07, 1.97688937e-01, 8.02310288e-01],
       [1.16231355e-07, 5.71946502e-02, 9.42805290e-01],
       [1.73847297e-06, 2.18725368e-01, 7.81272888e-01]], dtype=float32)

In [26]:
print(종속[-5:])

     품종_setosa  품종_versicolor  품종_virginica
145          0              0             1
146          0              0             1
147          0              0             1
148          0              0             1
149          0              0             1


In [28]:
print(독립.columns)

Index(['꽃잎길이', '꽃잎폭', '꽃받침길이', '꽃받침폭'], dtype='object')


In [27]:
# 학습한 가중치
model.get_weights()

[array([[ 1.3284669 ,  0.74916506, -0.5888545 ],
        [ 3.1035542 ,  0.31662115, -0.8744763 ],
        [-4.694901  , -1.0304421 ,  0.5870499 ],
        [-5.087944  , -1.9818534 ,  1.8619242 ]], dtype=float32),
 array([ 1.8505807,  1.0921242, -1.3351692], dtype=float32)]



*   가중치(Wx) 값과 bias들을 보여준다
*   (결과값의 소수점 3번째 자리까지 표기)

---

*   1열의 (1.328, 3,103, -4.694, -5.087)은 각각 품종_setosa에서 (꽃잎길이, 꽃잎폭, 꽃받침길이, 꽃받침폭)의 가중치 값을 나타낸다.
*   1열의 1.850은 품종_setosa의 bias값을 나타낸다.

---

*   2열, 3열의 경우에는 각각 품종_versicolor, 품종_virginica의 독립변수들의 가중치 값, bias 값을 나타낸다
